proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+---
+author: Thomas Bagrel
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/0>).
+**After creating the pull request, edit this file again, update the number in
+the link, and delete this bold sentence.**
+
+# Primitives to do destination-passing style structure building in compact regions
+
+This proposal introduces necessary tooling in GHC to enable direct data structure building in compact regions, in particular in destination-passing style (DPS), which can be useful both performance and expressivity purpose.
+
+## Motivation
+
+[Compact regions](https://github.com/ezyang/compact) have been supported in GHC since 2017; they provide a way to allocate memory in a region that can be garbage-collected all at once instead of being traversed fully by the GC at each collection cycle. Compact regions are useful in some performance-critical use-cases, especially when dealing with large long-lived datasets.
+
+So far however, there is no way to operate in a compact region directly. One has to first build data structures in the normal Haskell GC heap, and then copy them to a compact region.
+
+I propose a new compact region RTS primitive, [`compactAddHollow#`](https://gitlab.haskell.org/tweag/ghc/-/compare/master...tbagrel1%2Fdps-compact-regions-prims?from_project_id=1#451725cc4e5d443a3b7c2adcdf224840f953b7e2_3537_3537), that can be used to build data structures directly into a compact region, bypassing the need for a copy entirely. `compactAddHollow#` let us build structures in an _outside-in_ fashion, by allocating a specified constructor heap object in the compact region, without setting values for its fields yet.
+
+The main intent of `compactAddHollow#` is to be encapsulated in a destination-passing style API, where normal functions can receive _write references_, aka _destinations_, pointing to yet-uninitialized fields of the data structures in the compact region and use them to complete the structures.
+
+A real-world use case would be a program that parses a large serialized document and keep it in memory during the whole execution. Instead of being parsed into the GC heap, and maybe copied to a compact region afterwards, thanks to my change, the document can be parsed directly into the compact region, reducing greatly the time spent in garbage collection. I implemented a parser of S-expression in both naive style and DPS-style-for-compact-regions that are [presented in the **Examples** section](#library-side-performance-focused-example-s-expression-parser) The implementation that deserializes directly into the compact region (making use of the changes presented in the proposal) is up to twice as fast as the naive version thanks to garbage collection time saved.
```

Comment:
```suggestion
A real-world use case would be a program that parses a large serialized document and keep it in memory during the whole execution. Instead of being parsed into the GC heap, and maybe copied to a compact region afterwards, with this proposal, the document can be parsed directly into the compact region, reducing greatly the time spent in garbage collection. I implemented a parser of S-expression in both naive style and DPS-style-for-compact-regions that are [presented in the **Examples** section](#library-side-performance-focused-example-s-expression-parser) The implementation that deserializes directly into the compact region is up to twice as fast as the traditional version thanks to garbage collection time saved.
```

Note that if you just parse once at the beginning of the program, it doesn't account for much performance. A more load-intensive use-case has you parse a lot of documents but retain them for pretty well-defined amounts of time. Such as one-document-per-request on a web server or something like this.

Worth remembering that if the structure is retained for a while, you'll save some time because it's live data which isn't traversed. But you also save time on traversal of the structure at deallocation time (unless you're in a copying region of the GC, which you are the first time around, in which case dead data isn't traversed I believe).

A third way in which this avoids GC time is that nothing is allocated in the nursery. The nursery being full is what triggers a GC, so you expect to trigger GC less often.

I'm not sure it's worth commenting on all of this, but things to keep in mind. In short: we expect fewer, faster garbage collection phases.